File: e07/tests/precache.py
#!/usr/bin/env python3
"""Pre-render every fixed Russian reply with the real TeraTTS engine on the host → out/tts-cache/tera-cache/.
The APK ships these as assets/tera-cache/<key>.pcm; TeraTts.synthSentence serves them without synthesis."""
import os
from pathlib import Path
import re
import subprocess
import tempfile

root = Path(__file__).resolve().parents[1]
java_home = Path(os.environ.get("JAVA_HOME", "/usr/lib/jvm/java-26-openjdk"))
deps = Path(os.environ.get("E07_DEPENDENCIES", root / "out/dependencies"))
out = root / "out/tts-cache/tera-cache"
ort = deps / "verification/onnxruntime-1.17.1.jar"
# One pass over strings AND comments, so a phrase quoted inside a comment never reaches the renderer.
token = re.compile(r'"(?:\\.|[^"\\\n])*"|//[^\n]*|/\*.*?\*/', re.S)
reply = re.compile(r'^"([А-ЯЁ][^"\\]*)"$')   # reply literals start with a capital; keyword stems are lowercase
phrases = set()
for src in [*sorted((root / "src/com/stand/bridge").glob("*.java")), *sorted((root / "src/com/stand/tts").glob("*.java"))]:
    for match in token.finditer(src.read_text(encoding="utf-8")):
        found = reply.match(match.group())
        # Concatenation prefixes ("Заряд батареи: ", "Температура ") are never spoken on their own: they
        # keep the trailing space or punctuation that joins them to a runtime value.
        if found and found.group(1) == found.group(1).strip() and not found.group(1).endswith((":", ",")):
            phrases.add(found.group(1))
with tempfile.TemporaryDirectory(prefix="e07-precache-") as temporary:
    build = Path(temporary)
    tera = (root / "src/com/stand/tts/TeraTts.java").read_text(encoding="utf-8")
    pure = tera[tera.index("    public static String ttsNormalize"):tera.index("    private TeraTts()")]
    (build / "TeraTts.java").write_text("package com.stand.tts; public final class TeraTts {\n" + pure + "}\n", encoding="utf-8")
    (build / "phrases.txt").write_text("\n".join(sorted(phrases)) + "\n", encoding="utf-8")
    subprocess.run([str(java_home / "bin/javac"), "-encoding", "UTF-8", "-cp", str(ort), "-d", str(build),
                    str(build / "TeraTts.java"), *map(str, sorted((root / "src/com/stand/tts/tera").glob("*.java"))),
                    str(root / "src/com/stand/bridge/CommandHelp.java"), str(root / "src/com/stand/bridge/Ru2Zh.java"),
                    str(root / "tests/Precache.java")], check=True)
    subprocess.run([str(java_home / "bin/java"), "--enable-native-access=ALL-UNNAMED", "-cp", f"{build}{os.pathsep}{ort}",
                    "com.stand.bridge.Precache", str(deps), str(build / "phrases.txt"), str(out)], check=True)

File: e07/tests/Precache.java
package com.stand.bridge;

import com.stand.tts.TeraTts;
import com.stand.tts.tera.TeraTTS;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.TreeMap;

/** Host pre-renderer: every fixed Russian reply sentence → <out>/<key>.pcm (24 kHz mono PCM16), same engine,
 *  same accent lexicon as the car. Composed replies (help topics, trunk acknowledgements, seat greetings)
 *  are generated here because they never appear as one string literal. */
public final class Precache {
    public static void main(String[] args) throws Exception {
        Path deps = Paths.get(args[0]), phrases = Paths.get(args[1]), out = Paths.get(args[2]);
        String accents = TeraTts.accentVersion(Files.size(deps.resolve("assets/tera/ruaccent.bin")));
        System.out.println("accent version " + accents);
        Files.createDirectories(out);
        LinkedHashSet<String> texts = new LinkedHashSet<>(Files.readAllLines(phrases));
        Ru2Zh.ALLOW_TRUNK = true; Ru2Zh.ALLOW_UNSAFE = false;
        for (int i = 0; i <= CommandHelp.TOPICS.length; i++) texts.add(CommandHelp.spoken(i));
        for (String action : new String[]{"открыть", "закрыть", "остановить"})
            for (String label : new String[]{"нижний борт", "крышу багажника", "заднее стекло", "стекло перегородки",
                    "весь задний отсек", "нижний борт и крышу багажника"})
                texts.add("Передала команду: " + action + " " + label);
        for (String seat : new String[]{"Пассажир", "Слева сзади", "Справа сзади", "По центру", "Сзади"})
            texts.add(seat + ", чем могу помочь");
        TreeMap<String, String> index = new TreeMap<>();
        int rendered = 0;
        try (TeraTTS engine = new TeraTTS(deps.resolve("assets/tera"), TeraTts.VOICE)) {
            for (String text : texts) {
                if (text.isBlank()) continue;
                for (String sentence : TeraTts.sentences(TeraTts.ttsNormalize(text))) {
                    String key = TeraTts.cacheKey(sentence, 24000, accents);
                    index.put(key, sentence);
                    Path file = out.resolve(key + ".pcm");
                    if (Files.exists(file)) continue;
                    byte[] pcm = TeraTts.pcm16(engine.synthesize(sentence, TeraTts.SPEED), TeraTts.NATIVE_RATE, 24000);
                    if (pcm.length < 2400) throw new IllegalStateException("suspiciously short render: " + sentence);
                    Files.write(file, pcm);
                    rendered++;
                }
            }
        }
        long stale = 0, bytes = 0;
        try (var files = Files.list(out)) {
            for (Path f : (Iterable<Path>) files::iterator) {
                String name = f.getFileName().toString();
                if (!name.endsWith(".pcm")) continue;
                if (!index.containsKey(name.substring(0, name.length() - 4))) { Files.delete(f); stale++; }
                else bytes += Files.size(f);
            }
        }
        StringBuilder sb = new StringBuilder();
        for (Map.Entry<String, String> e : index.entrySet()) sb.append(e.getKey()).append('\t').append(e.getValue()).append('\n');
        Files.writeString(out.resolveSibling("index.tsv"), sb.toString());
        System.out.println("PRECACHE texts=" + texts.size() + " sentences=" + index.size() + " rendered_now=" + rendered
                + " stale_removed=" + stale + " bytes=" + bytes);
    }
}
